File: step_definitions/Graceful Error Handling for Unavailable Tracking Information.java
package com.cucumber.steps;

import com.framework.cucumber.TestHarness;
import io.cucumber.java.en.Given;
import io.cucumber.java.en.Then;
import io.cucumber.java.en.When;
import com.page_objects.TrackingPage;

public class TrackingSteps extends TrackingPage {

    private TestHarness testHarness = new TestHarness();

    @Given("the user has access to the tracking feature with User ID {string}")
    public void theUserHasAccessToTheTrackingFeatureWithUserID(String userId) {
        navigateToTrackingFeature(userId);
    }

    @When("the user navigates to the tracking page")
    public void theUserNavigatesToTheTrackingPage() {
        openTrackingPage();
    }

    @Then("the tracking page is displayed")
    public void theTrackingPageIsDisplayed() {
        verifyTrackingPageDisplayed();
    }

    @Given("the user is on the tracking page")
    public void theUserIsOnTheTrackingPage() {
        openTrackingPage();
    }

    @When("the user enters an invalid tracking number {string}")
    public void theUserEntersAnInvalidTrackingNumber(String trackingNumber) {
        enterTrackingNumber(trackingNumber);
    }

    @Then("the system displays an error message indicating tracking information is unavailable")
    public void theSystemDisplaysAnErrorMessageIndicatingTrackingInformationIsUnavailable() {
        verifyErrorMessageForUnavailableTracking();
    }

    @Given("the user has entered an invalid tracking number {string}")
    public void theUserHasEnteredAnInvalidTrackingNumber(String trackingNumber) {
        enterTrackingNumber(trackingNumber);
    }

    @When("the error message is displayed")
    public void theErrorMessageIsDisplayed() {
        verifyErrorMessageDisplayed();
    }

    @Then("the error message is clear and suggests alternative actions")
    public void theErrorMessageIsClearAndSuggestsAlternativeActions() {
        verifyErrorMessageClarityAndGuidance();
    }

    @Given("the user is on the tracking page with an error message displayed")
    public void theUserIsOnTheTrackingPageWithAnErrorMessageDisplayed() {
        openTrackingPageWithErrorMessage();
    }

    @When("the user attempts to refresh the tracking page")
    public void theUserAttemptsToRefreshTheTrackingPage() {
        refreshTrackingPage();
    }

    @Then("the system handles the refresh gracefully without crashing")
    public void theSystemHandlesTheRefreshGracefullyWithoutCrashing() {
        verifyGracefulRefreshHandling();
    }

    @Given("an error message is displayed for unavailable tracking information")
    public void anErrorMessageIsDisplayedForUnavailableTrackingInformation() {
        verifyErrorMessageForUnavailableTracking();
    }

    @When("the error occurs")
    public void theErrorOccurs() {
        simulateErrorOccurrence();
    }

    @Then("the error is logged in the system logs")
    public void theErrorIsLoggedInTheSystemLogs() {
        verifyErrorLogging();
    }

    @Given("the user has entered an invalid tracking number")
    public void theUserHasEnteredAnInvalidTrackingNumber() {
        enterInvalidTrackingNumber();
    }

    @When("the user decides to retry entering a tracking number")
    public void theUserDecidesToRetryEnteringATrackingNumber() {
        retryEnteringTrackingNumber();
    }

    @Then("the user is allowed to retry entering a tracking number")
    public void theUserIsAllowedToRetryEnteringATrackingNumber() {
        verifyRetryAllowed();
    }

    @When("the user enters a valid tracking number")
    public void theUserEntersAValidTrackingNumber() {
        enterValidTrackingNumber();
    }

    @Then("tracking information is displayed correctly")
    public void trackingInformationIsDisplayedCorrectly() {
        verifyTrackingInformationDisplayed();
    }

    @Given("the user is retrieving tracking information")
    public void theUserIsRetrievingTrackingInformation() {
        startTrackingInformationRetrieval();
    }

    @When("a network failure occurs")
    public void aNetworkFailureOccurs() {
        simulateNetworkFailure();
    }

    @Then("the system displays a network error message and suggests retrying")
    public void theSystemDisplaysANetworkErrorMessageAndSuggestsRetrying() {
        verifyNetworkErrorMessageAndSuggestion();
    }

    @Given("the tracking service is down")
    public void theTrackingServiceIsDown() {
        simulateTrackingServiceDown();
    }

    @When("the user attempts to retrieve tracking information")
    public void theUserAttemptsToRetrieveTrackingInformation() {
        attemptTrackingInformationRetrieval();
    }

    @Then("the system displays a service unavailable message")
    public void theSystemDisplaysAServiceUnavailableMessage() {
        verifyServiceUnavailableMessage();
    }

    @When("the user reads the error message")
    public void theUserReadsTheErrorMessage() {
        readErrorMessage();
    }

    @Then("contact support information is provided in the error message")
    public void contactSupportInformationIsProvidedInTheErrorMessage() {
        verifyContactSupportInformationInErrorMessage();
    }

    @Given("the system is under high load")
    public void theSystemIsUnderHighLoad() {
        simulateHighLoadScenario();
    }

    @When("the user attempts to access tracking information")
    public void theUserAttemptsToAccessTrackingInformation() {
        attemptAccessingTrackingInformation();
    }

    @Then("system performance remains stable and handles errors gracefully")
    public void systemPerformanceRemainsStableAndHandlesErrorsGracefully() {
        verifySystemPerformanceAndErrorHandling();
    }

    @Given("the tracking service was down")
    public void theTrackingServiceWasDown() {
        simulateTrackingServiceDown();
    }

    @When("the tracking service resumes")
    public void theTrackingServiceResumes() {
        simulateTrackingServiceResumption();
    }

    @Then("tracking information is displayed correctly once the service resumes")
    public void trackingInformationIsDisplayedCorrectlyOnceTheServiceResumes() {
        verifyTrackingInformationAfterServiceResumption();
    }

    @Given("the tracking service is currently unavailable")
    public void theTrackingServiceIsCurrentlyUnavailable() {
        simulateTrackingServiceUnavailability();
    }

    @Then("estimated recovery time is displayed in the error message")
    public void estimatedRecoveryTimeIsDisplayedInTheErrorMessage() {
        verifyEstimatedRecoveryTimeInErrorMessage();
    }

    @Given("the user enters an invalid tracking format")
    public void theUserEntersAnInvalidTrackingFormat() {
        enterInvalidTrackingFormat();
    }

    @When("the tracking number is submitted")
    public void theTrackingNumberIsSubmitted() {
        submitTrackingNumber();
    }

    @Then("the system displays an error message for invalid tracking formats")
    public void theSystemDisplaysAnErrorMessageForInvalidTrackingFormats() {
        verifyErrorMessageForInvalidTrackingFormat();
    }

    @Given("the user has experienced an error")
    public void theUserHasExperiencedAnError() {
        simulateUserErrorExperience();
    }

    @Then("the user is prompted to provide feedback on the error handling experience")
    public void theUserIsPromptedToProvideFeedbackOnTheErrorHandlingExperience() {
        verifyFeedbackPromptForErrorHandling();
    }
}

File: step_definitions/Graceful Error Handling for Unavailable Tracking Information.js
import { Given, When, Then } from "cypress-cucumber-preprocessor/steps";
import TrackingPage from "../pages/trackingPage";

Given("the user has access to the tracking feature with User ID {string}", (userId) => {
  // Assume user session is set up with the given User ID
  cy.setUserSession(userId);
});

When("the user navigates to the tracking page", () => {
  TrackingPage.visit();
});

Then("the tracking page is displayed", () => {
  TrackingPage.verifyPageDisplayed();
});

Given("the user is on the tracking page", () => {
  TrackingPage.verifyPageDisplayed();
});

When("the user enters an invalid tracking number {string}", (trackingNumber) => {
  TrackingPage.enterTrackingNumber(trackingNumber);
});

Then("the system displays an error message indicating tracking information is unavailable", () => {
  TrackingPage.verifyErrorMessage("Tracking information is unavailable");
});

Given("the user has entered an invalid tracking number {string}", (trackingNumber) => {
  TrackingPage.enterTrackingNumber(trackingNumber);
});

When("the error message is displayed", () => {
  TrackingPage.verifyErrorMessageDisplayed();
});

Then("the error message is clear and suggests alternative actions", () => {
  TrackingPage.verifyErrorMessageContent("Please check the tracking number and try again.");
});

Given("the user is on the tracking page with an error message displayed", () => {
  TrackingPage.verifyPageDisplayed();
  TrackingPage.verifyErrorMessageDisplayed();
});

When("the user attempts to refresh the tracking page", () => {
  cy.reload();
});

Then("the system handles the refresh gracefully without crashing", () => {
  TrackingPage.verifyPageDisplayed();
});

Given("an error message is displayed for unavailable tracking information", () => {
  TrackingPage.verifyErrorMessageDisplayed();
});

When("the error occurs", () => {
  // Simulate error occurrence
});

Then("the error is logged in the system logs", () => {
  cy.verifyLogEntry("Tracking information unavailable error");
});

Given("the user has entered an invalid tracking number", () => {
  TrackingPage.enterTrackingNumber("00000");
});

When("the user decides to retry entering a tracking number", () => {
  TrackingPage.clearTrackingNumberField();
});

Then("the user is allowed to retry entering a tracking number", () => {
  TrackingPage.enterTrackingNumber("12345");
  TrackingPage.verifyTrackingInformationDisplayed();
});

Given("the user is retrieving tracking information", () => {
  TrackingPage.enterTrackingNumber("12345");
});

When("a network failure occurs", () => {
  cy.simulateNetworkFailure();
});

Then("the system displays a network error message and suggests retrying", () => {
  TrackingPage.verifyNetworkErrorMessage("Network error occurred. Please try again.");
});

Given("the tracking service is down", () => {
  cy.simulateServiceDown();
});

When("the user attempts to retrieve tracking information", () => {
  TrackingPage.enterTrackingNumber("12345");
});

Then("the system displays a service unavailable message", () => {
  TrackingPage.verifyServiceUnavailableMessage("Service is currently unavailable.");
});

Given("the system is under high load", () => {
  cy.simulateHighLoad();
});

When("the user attempts to access tracking information", () => {
  TrackingPage.enterTrackingNumber("12345");
});

Then("system performance remains stable and handles errors gracefully", () => {
  TrackingPage.verifyTrackingInformationDisplayed();
});

Given("the tracking service was down", () => {
  cy.simulateServiceDown();
});

When("the tracking service resumes", () => {
  cy.simulateServiceResume();
});

Then("tracking information is displayed correctly once the service resumes", () => {
  TrackingPage.verifyTrackingInformationDisplayed();
});

Given("the tracking service is currently unavailable", () => {
  cy.simulateServiceDown();
});

When("the user reads the error message", () => {
  TrackingPage.verifyErrorMessageDisplayed();
});

Then("estimated recovery time is displayed in the error message", () => {
  TrackingPage.verifyErrorMessageContent("Estimated recovery time: 2 hours");
});

Given("the user enters an invalid tracking format", () => {
  TrackingPage.enterTrackingNumber("invalidFormat");
});

When("the tracking number is submitted", () => {
  TrackingPage.submitTrackingNumber();
});

Then("the system displays an error message for invalid tracking formats", () => {
  TrackingPage.verifyErrorMessage("Invalid tracking format");
});

Given("the user has experienced an error", () => {
  TrackingPage.verifyErrorMessageDisplayed();
});

When("the error message is displayed", () => {
  TrackingPage.verifyErrorMessageDisplayed();
});

Then("the user is prompted to provide feedback on the error handling experience", () => {
  TrackingPage.verifyFeedbackPromptDisplayed();
});